### DISCUSSION TITLE
Glass opacity

### DISCUSSION THREAD

**Discussion by Shmew** [jasonjgardner/jg-rtx#18] [Feedback]
Firstly I want to say that I love this texture pack and I'm appreciative of the work you're doing!

The biggest issue I've run into with this pack is the glass. It's very difficult to see through.

RTX enabled:
![image](https://user-images.githubusercontent.com/11186595/111892601-f1ea5f00-89ca-11eb-92fc-3cd3f8476925.png)

RTX disabled:
![image](https://user-images.githubusercontent.com/11186595/111892610-fe6eb780-89ca-11eb-9e95-bd8c10512a2c.png)

I also personally dislike the wood frames going through the block but that's less of an issue for me.

**Comment by jasonjgardner** [comment]
Thanks for the feedback. I agree! I thought the middle frames would spice things up at first, but they don't look good when you've got a bunch of windows together, or over the beacon. I think a single pane for the normal glass might be best.

As for the window transparency, I've had a hard time figuring out the best opacity and reflectivity for a few reasons. The window renders in Substance Painter look slightly different than how they appear in Minecraft. The stained glass windows need to be more opaque in order to make the RTX light coming in more colorful; however, the opacity still needs to stay within a certain threshold because the [PBR texture guide says](https://www.nvidia.com/en-us/geforce/guides/minecraft-rtx-texturing-guide/#textcomponentenhance_2056881019f5559ecc_7d59_4cdf_94df_8ea4b640b7f8):

>Note: In the initial release of Minecraft with RTX, the client will use the same textures for both RTX enabled and disabled. This means that an opacity value that is tuned for RTX enabled might look too transparent when RTX is disabled.  To help balance this out, the game will only use the upper half of the range of the texture’s alpha channel (0.5 - 1) when RTX is enabled, but it will map this range to the full range of opacity in the game so no functionality is lost. For RTX disabled, the full range of the alpha channel will be mapped to the full range opacity in the game.

I'll try something different. I'm glad you said something!

**Comment by jasonjgardner** [comment]
It seems like the problem was caused by the file format. Converting the glass texture from PNG to TGA seems to have solved the problem. (A PNG is still currently intentionally included in the [latest release's .mcpack.](https://github.com/jasonjgardner/jg-rtx/releases/download/0.15.1-alpha/JG-RTX_0.15.1.mcpack))
> ![PNG vs TGA](https://user-images.githubusercontent.com/1903667/112560964-3ea2b100-8da2-11eb-9b37-20e52950e462.jpg)
> PNG glass vs. TGA glass

> ![Glass with RTX on and off](https://user-images.githubusercontent.com/1903667/112560979-482c1900-8da2-11eb-927f-e190857ab813.jpg)
> New glass texture: RTX on/off

**Reply by erich666** [reply]
So strange! And I apologize for misleading you and talking you into PNGs. They really shouldn't be different. What I was told was that any difference occurs because of poor TGA to PNG conversion by some tools. Reading more about TGA today, ultimately, I think the problem might be that there's a confusion with TGA having to do with associated (premultiplied) vs. unassociated alphas. [This post](https://newsgroup.xnview.com/viewtopic.php?t=27812) notes it, though the problem's very old and I would have thought it would have been resolved by now. TGA once supported both associated and unassociated alphas, but some tools do not do a good job of paying attention. So, what tool do you use to convert from TGA to PNG, or vice versa? I'm fine with you sticking to TGA - it works! - but I'd like to understand the root causes here better. My tools use PNG, so I want to know where the problem is coming from and how I should avoid it.

You say "A PNG is still currently intentionally included in the latest release's .mcpack" but I'm not 100% sure where to go with that information (which PNG, for starters? And how can I compare it to the TGA version?). Can you provide a link to the PNG mcpack link and the TGA mcpack link, exactly? Giving an old pack is fine.  Or, can you exactly point to "this TGA works, this PNG doesn't" and I'll form my own .mcpack (wish me luck - I've never tried that before).

Thanks for any help - I definitely want to dig in on this one, for my own knowledge.

**Reply by jasonjgardner** [reply]
Seems like PNGs are still a better choice for the textures which do not need to be semi-transparent. Plant textures are fine as PNGs, for example. Glass and water textures might need to be TGAs to get the correct opacity.

I'm using Substance Painter to output the glass material as either a PNG or a TGA. I've also used Photoshop to alter the levels of a TGA's alpha channel, though I didn't have to do it with the glass block. 

Here are the glass files:
https://github.com/jasonjgardner/jg-rtx/blob/c82da70c0aa14d8a1b2f631113ea0d6767a35631/development_resource_packs/JG-RTX/textures/blocks/glass.tga
https://github.com/jasonjgardner/jg-rtx/blob/c82da70c0aa14d8a1b2f631113ea0d6767a35631/development_resource_packs/JG-RTX/textures/blocks/glass.png

.mcpack files are zip archives, so you can create them or extract their contents just by changing the file extension.

**Reply by erich666** [reply]
Thanks, and it's strange, these two files are quite different in their contents, in almost every way! Are you sure these two files are from the same material description? If so, then I suspect the tools are doing something "interesting" when you save as a TGA vs. a PNG. Looking at the values, TGA vs. PNG (using [Beyond Compare](https://www.scootersoftware.com/download.php), which is great for this sort of thing and has a 30 day free trial):

![image](https://user-images.githubusercontent.com/803995/112883464-53ca5900-909c-11eb-859e-22be9e9d550a.png)

The alpha is much lower (i.e., more transparent) for the TGA image: 24 vs. 75, that's three times as transparent. That's probably most of the problem right there. The numbers in the semi-transparent areas are also considerably different: 188,201,209 vs. 121,121,121. I don't know where it's getting those 121 values - first, they're all the same; second, they're not premultiplied values (which would be lower than unassociated values). Looking around at other texels in the PNG, they're all grayscale values, while in the TGA they're all colors.

So, yeah, I'm thinking the problem is the tool, not the image file format. It seems to be turning the semitransparent colors gray and darker, and, more importantly, considerably changing the alpha values.

Could you do me a favor and try one little experiment? Read a semi-transparent TGA into [paint.net](https://www.getpaint.net/) (my other go-to utility, Irfanview, is faulty when writing out TGA's). Save as a PNG. Try using that PNG. Does it look OK in MinecraftRTX compared to the TGA? I think it should (but my MinecraftRTX is kind of trashed right now, so I hope you can check). If it doesn't, I _definitely_ want to know that.

I'll attach mine here - you should get the same result.

I'm of course fine with you sticking with TGA - no reason to change your workflow based on my needs. Though, if you find out PNGs do work fine, you might consider using them, since they're a lot smaller in size and better supported by most utilities.

Here's my PNG version, saved out of paint.net, of your glass.tga file:
![glass_was_tga](https://user-images.githubusercontent.com/803995/112885050-42824c00-909e-11eb-8e80-28d4cb1dc446.png)

**Comment by art926** [comment]
Was the issue fixed in 0.15.1-alpha ? In the description for that version it says "Converted glass texture file from PNG to TGA to improve glass transparency", but when I update the pack and try it in the Hilltop Lifestyle demo world, the lighthouse windows are still very opaque.

**Reply by jasonjgardner** [reply]
This was fixed for just the normal glass texture. The stained glass textures are still PNG files and have this transparency issue, [but those will get fixed soon](https://github.com/jasonjgardner/jg-rtx/issues/23).

Does the smooth quartz stair texture work in your Hilltop Lifestyle world? It's missing when I play...
